src/Engines/DatabaseEngine.php

```diff
@@ -0,0 +1,273 @@
+     *
+     * @param  \Illuminate\Database\Eloquent\Collection  $models
+     * @return void
+     */
+    public function delete($models)
+    {
+        //
+    }
+
+    /**
+     * Perform the given search on the engine.
+     *
+     * @param  \Laravel\Scout\Builder  $builder
+     * @return mixed
+     */
+    public function search(Builder $builder)
+    {
+        $models = $this->searchModels($builder);
+
+        return [
+            'results' => $models->all(),
+            'total' => count($models),
+        ];
+    }
+
+    /**
+     * Perform the given search on the engine.
+     *
+     * @param  \Laravel\Scout\Builder  $builder
+     * @param  int  $perPage
+     * @param  int  $page
+     * @return mixed
+     */
+    public function paginate(Builder $builder, $perPage, $page)
+    {
+        $models = $this->searchModels($builder);
+
+        return [
+            'results' => $models->forPage($page, $perPage)->all(),
+            'total' => count($models),
+        ];
+    }
+
+    /**
+     * Get the Eloquent models for the given builder.
+     *
+     * @param  \Laravel\Scout\Builder  $builder
+     * @return \Illuminate\Database\Eloquent\Collection
+     */
+    protected function searchModels(Builder $builder)
+    {
+        $columns = array_keys($builder->model->toSearchableArray());
+
+        $query = $builder->model->query()
+                        ->whereFulltext($columns, $builder->query)
+                        ->when(! is_null($builder->callback), function ($query) use ($builder) {
+                            call_user_func($builder->callback, $query, $builder, $builder->query);
+                        })
+                        ->when(! $builder->callback && count($builder->wheres) > 0, function ($query) use ($builder) {
+                            foreach ($builder->wheres as $key => $value) {
+                                if ($key !== '__soft_deleted') {
+                                    $query->where($key, $value);
+                                }
+                            }
+                        })
+                        ->when(! $builder->callback && count($builder->whereIns) > 0, function ($query) use ($builder) {
+                            foreach ($builder->whereIns as $key => $values) {
+                                $query->whereIn($key, $values);
+                            }
+                        })
+                        ->orderBy($builder->model->getKeyName(), 'desc');
+
+        $models = $this->ensureSoftDeletesAreHandled($builder, $query)
+                        ->get()
+                        ->values();
+
+        if (count($models) === 0) {
+            return $models;
+        }
+
+        return $models->filter(function ($model) use ($builder) {
+            if (! $model->shouldBeSearchable()) {
+                return false;
+            }
+
+            if (! $builder->query) {
+                return true;
+            }
+
+            return false;
```

Comment:
I feel like I must be missing something really obvious here. This `filter` operation will basically always remove every model from the result set unless the search query is empty? How does this ever actually return any results in normal circumstances?

Reply:
Odd, it worked when I tried it but your remarks seem to make sense yeah.

Reply:
I pushed a change for this. I think I must have reverted something the last time I pushed. Anyway, it's fixed now.